### DISCUSSION TITLE
浅谈 Go 语言 select 的实现原理 · /golang-select

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#559] [Announcements]
https://draveness.me/golang-select

**Comment by arkrz** [comment]
您好，总结中说：
```
当调度器唤醒当前 Goroutine 时就会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 sudog 结构并返回对应的索引；
```
那么是不是意味着，唤醒该G的通道不一定会被执行收发操作？

**Comment by draveness** [comment]
> @ThomasZN
> 您好，总结中说：
> ```
> 当调度器唤醒当前 Goroutine 时就会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 sudog 结构并返回对应的索引；
> ```
> 那么是不是意味着，唤醒该G的通道不一定会被执行收发操作？

没太明白你想问啥...

**Comment by arkrz** [comment]
> @draveness
> > @ThomasZN
> > 您好，总结中说：
> > ```
> > 当调度器唤醒当前 Goroutine 时就会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 sudog 结构并返回对应的索引；
> > ```
> > 那么是不是意味着，唤醒该G的通道不一定会被执行收发操作？
> 
> 没太明白你想问啥...

好吧，原文中说清楚了，是我疏忽了:(

**Comment by draveness** [comment]
@ThomasZN 没事

**Comment by SmileEye** [comment]
写的真好

**Comment by SmileEye** [comment]
大佬您好，有个小问题，文中提到 空select{} 会引起阻塞，或者当select语句中只有一个case且case中处理的channel为nil时，会调用block() 造成阻塞协程休眠，这两个都试了下，都提示死锁，进程退出，是不是我代码哪里写的有问题，您看下：

1.

```
func main() {
	select{}
}
```

运行提示：fatal error: all goroutines are asleep - deadlock!

2. 

```
func main() {
	var ch chan int
	select {
		case <-ch:
	}
}
```

运行提示：fatal error: all goroutines are asleep - deadlock!

**Comment by draveness** [comment]
@SmileEye  Go 语言死锁检测会发现当前的 Goroutine 已经不可能被唤醒，就会直接报错退出，之后会澄清一下这个问题，避免造成更多的疑惑

**Comment by SmileEye** [comment]
> @draveness
> @SmileEye  Go 语言死锁检测会发现当前的 Goroutine 已经不可能被唤醒，就会直接报错退出，之后会澄清一下这个问题，避免造成更多的疑惑

嗯嗯，谢谢大佬解惑

**Comment by wwulfric** [comment]
现在画图怎么不上色了😂

**Comment by tricleoo** [comment]
您好，假设select时从不同的chan中接收数据，select处于阻塞，多核情况下同时收到数据，会不会导致数据丢失？看代码协程挂起，被唤醒后加锁，然后取了一个数据；加锁前会不会出现多个sudog收到数据？

**Comment by tricleoo** [comment]
我又看了下代码，dequeue时有判断g.selectDone的标志，应该是这个防止竞争的吧

**Comment by Hawken94** [comment]
您好，在 *循环* --> 'caseSend'的那一段中 `如果当前 Channel 已经被关闭就会直接跳到 rclose 代码段` ，应该是直接跳到 sclose 代码段吧。

**Comment by draveness** [comment]
> 您好，在 _循环_ --> 'caseSend'的那一段中 `如果当前 Channel 已经被关闭就会直接跳到 rclose 代码段` ，应该是直接跳到 sclose 代码段吧。

看得很仔细..确实有问题，已经改了

**Comment by zzz6519003** [comment]
select的本质&作用？

**Comment by MrYueQ** [comment]
[总结](https://draveness.me/golang/keyword/golang-select.html#%E6%80%BB%E7%BB%93) 

第4点，
如果 select 语句中只包含两个 case 并且其中一个是 default，那么 Channel ***和***接收和发送操作都会使用 selectnbrecv 和 selectnbsend 非阻塞地执行接收和发送操作；

**Comment by draveness** [comment]
@MrYueQ 修复了

**Comment by aizk** [comment]
如果 OCASE 节点的都是空的，就意味着这是一个 default 节点:

这里是不是少一个词，节点的？？？都是空的。

节点的数据都是空的？还是节点的指针都是空的？

**Comment by draveness** [comment]
> 如果 OCASE 节点的都是空的，就意味着这是一个 default 节点:
> 
> 这里是不是少一个词，节点的？？？都是空的。
> 
> 节点的数据都是空的？还是节点的指针都是空的？

节点的条件，已经更新了

**Comment by bluntdel** [comment]
> @ThomasZN
> 您好，总结中说：
> ```
> 当调度器唤醒当前 Goroutine 时就会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 sudog 结构并返回对应的索引；
> ```
> 那么是不是意味着，唤醒该G的通道不一定会被执行收发操作？
一定会执行吧？毕竟是在Channel的等待队列里。然后会返回对应的case序号，处理sudog。

**Comment by yanjinbin** [comment]
Ω  这也太强了

**Comment by yanjinbin** [comment]
希望能说说写这篇文章的过程 学习学习

**Comment by bolenzhang** [comment]
lockorder的作用还是没太理解透，哪位大神能否举个实例帮忙解释下？

**Comment by bolenzhang** [comment]
> 当调度器唤醒当前 Goroutine 时就会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 sudog 结构并返回对应的索引；

为啥需要用lockorder遍历所有case，不是有pollorder查看是否有被激活的channel就够了吗？

**Comment by draveness** [comment]
> lockorder的作用还是没太理解透，哪位大神能否举个实例帮忙解释下？

> > 当调度器唤醒当前 Goroutine 时就会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 sudog 结构并返回对应的索引；
> 
> 为啥需要用lockorder遍历所有case，不是有pollorder查看是否有被激活的channel就够了吗？

lockOrder 的主要作用是避免死锁，如果两个 Goroutine 都需要锁定 ChannelA 和 ChannelB 才能执行任务，当两者尝试去依照不同的顺序进行锁定时，就可能发生死锁，以下是同时发生的：

1. Goroutine1 先锁定 A，这时发现 B 被锁定了，它是会有 A 的锁，等待 B 的释放
1. Goroutine2 先锁定 B，这时发现 A 被锁定了，它会持有 B 的锁，等待 A 的释放

如果锁定的顺序相同，这种情况就不会出现了，Goroutine1 和 2 都按照字母序来锁定 Channel，先获得 A 的 Goroutine 就可以先执行

**Reply by dablelv** [reply]
为什么会有两个 Goroutine 锁定 Channel，channel 对其并发读写时不是不需要再显示加锁了吗？

**Comment by rikewang** [comment]
> 在select中多个case同时触发，会随机执行一个case，但是golang为什么这样设计，顺序的不行吗？

大概想了下，如果每次都是按顺序，可能剩下不会得到执行。

**Comment by draveness** [comment]
> 在select中多个case同时触发，会随机执行一个case，但是golang为什么这样设计，顺序的不行吗？

我看了远古时代的第一版 select 就是这样的，commit 里并没有说明为什么要这么干。

https://github.com/golang/go/commit/cb9b1038db77198c2b0961634cf161258af2374d

这个做法的目的应该是为了保证公平，要不然多个 Channel 按顺序执行就会有饥饿问题

**Comment by rikewang** [comment]
是的，可能会饥饿问题。感谢恢复好及时~

**Comment by tidyoux** [comment]
> 3. caseSend：当前 case 会向 Channel 发送数据；

漏掉了 `bufsend ` 的情况，但是对应的 `图 5-8` 是正确的。

**Comment by bzd111** [comment]
我们只是向看看 Channel 的可读或者可写状态 => 想
select 存在量个 case => 两

**Comment by bloatfan** [comment]
> 我又看了下代码，dequeue时有判断g.selectDone的标志，应该是这个防止竞争的吧

很棒，通过 cas 来保证，这个 g，只能被唤醒一次  👍

**Comment by qiluge** [comment]
请问包含多个chan和一个default case的情况下是怎么执行的？我试了下还是非阻塞的。

**Comment by draveness** [comment]
> 请问包含多个chan和一个default case的情况下是怎么执行的？我试了下还是非阻塞的。

只要包含 default 就是非阻塞的

**Comment by ikingye** [comment]
`errCh := make(chan error, len(tasks)` 后面少了括号

---

*2020-06-05 UPDATES: 已修复*

**Comment by shenbaise9527** [comment]
* 如果当前 Channel 的 sendq 上有等待的 Goroutine，就会跳到 recv 标签从 Goroutine 中读取数据；
* 如果当前 Channel 的缓冲区不为空，就会跳到 bufrecv 标签处从缓冲区获取数据；
* 如果当前 Channel 已经被关闭，就会跳到 rclose 做一些清除的收尾工作；


第一个描述觉得不太准确，如果channel为带缓冲区的，进入到recv标签里，select分支获取到的数据也是从缓冲区读取的，但同时会从等待的Goroutine中读取数据放入到缓冲区相同的位置上。描述应该更详细点，以免误解了

---

*2020-07-04 UPDATES: 已修复*

**Comment by linvon** [comment]
作者是否有考虑调整一下章节间的顺序？本文中提到了 channel 章节中相关的内容，而且逻辑上感觉先介绍 channel 更容易理解一些，但目前 channel 章节是在更后面的章节里的。

**Comment by draveness** [comment]
> 作者是否有考虑调整一下章节间的顺序？本文中提到了 channel 章节中相关的内容，而且逻辑上感觉先介绍 channel 更容易理解一些，但目前 channel 章节是在更后面的章节里的。

可能会调整，但是这里应该不会，大多数的只是都是网状的会相互依赖，要用线性的文章来表示网状的知识本来就可能出现这种问题

**Comment by Formergg** [comment]
typo：`提交将 x, ok := <-c 语法删除删除`   结尾`删除`重复了

---

*2020-09-07 UPDATES: 已修复*

**Comment by woofyzhao** [comment]
1. 从nil channel收发就永久阻塞, 为什么这么设计?
2. 通过sudog串联来和lock order对应判断唤醒的case的index这个挺有意思的.

**Comment by draveness** [comment]
> 1. 从nil channel收发就永久阻塞, 为什么这么设计?

个人认为从 nil channel 收发在设计上应该有有两个相对合理的选择：崩溃和阻塞，但是如果使用 select 监听多个 Channel 其中只有一个是 nil 并不会永久阻塞，所以我觉得这个选择也算得上是合理：

```go
package main

import (
	"fmt"
)

func main() {
	var fooCh chan int
	barCh := make(chan int)
	
	go func() {
		barCh <- 1
	}()
	select {
	case <-fooCh:
	case <-barCh:
		fmt.Println("bar")
	}
}
```

**Comment by qichengzx** [comment]
"select 会选择那个 case 执行的问题"
那->哪

---

*2020-12-08 UPDATES: 已修复*

**Comment by jackzhou121** [comment]
看到runtime.selectgo初始化的部分需要确定pollorder和lockorder，有点疑惑在多个channel处于ready的时候，select会随机选择一个ready channel处理，所以pollorder不是运行时才能确定吗？为什么在初始化阶段会确定pollorder？

**Comment by jackzhou121** [comment]
lockorder有些疑惑： lock操作对象应该是channel，如果有多个channel ready，假设 goroutine-A 处理channel-1,  goroutine-B 是否可以处理channel-2 ？ 还是要等到goroutine-A 处理完所有的channel之后，其他goroutine才能处理

**Comment by Northward712** [comment]
> lockorder有些疑惑： lock操作对象应该是channel，如果有多个channel ready，假设 goroutine-A 处理channel-1, goroutine-B 是否可以处理channel-2 ？ 还是要等到goroutine-A 处理完所有的channel之后，其他goroutine才能处理

怎么说呢。。
当前的goroutine只管当前的select状态，跟其他goroutine的没啥关系
应该是这么理解

**Comment by Northward712** [comment]
> 看到runtime.selectgo初始化的部分需要确定pollorder和lockorder，有点疑惑在多个channel处于ready的时候，select会随机选择一个ready channel处理，所以pollorder不是运行时才能确定吗？为什么在初始化阶段会确定pollorder？

初始化阶段就是运行期间调用runtime.selectgo
只不过是在编译期把一个包含三个 case 的正常 select 语句其实会被展开成了那种形式。

**Comment by sunnystitch** [comment]
如果当前的 select 条件只包含一个 case，那么编译器会将 select 改写成 if 条件语句。下面对比了了改写前后的代码：    对比了了    发现一个笔误  


---

*2021-01-24 UPDATES: 已修复*

**Comment by dillanzhou** [comment]
> @jackzhou121
> lockorder有些疑惑： lock操作对象应该是channel，如果有多个channel ready，假设 goroutine-A 处理channel-1,  goroutine-B 是否可以处理channel-2 ？ 还是要等到goroutine-A 处理完所有的channel之后，其他goroutine才能处理

这里的一个疑问是为什么selectgo中要用sellock一次性把所有chan都锁住，为啥不是访问哪个锁哪个。golang团队在2014年https://github.com/golang/go/issues/8896 尝试开发过细粒度的锁，性能会有明显提升，但是无法保证这个实现在所有情况下都有和当前实现相同的表现。所以这个尝试最终被放弃了。。。

**Comment by dillanzhou** [comment]
> @draveness
> > 1. 从nil channel收发就永久阻塞, 为什么这么设计?
> 
> 个人认为从 nil channel 收发在设计上应该有有两个相对合理的选择：崩溃和阻塞，但是如果使用 select 监听多个 Channel 其中只有一个是 nil 并不会永久阻塞，所以我觉得这个选择也算得上是合理：
> 
> ```go
> package main
> 
> import (
> 	"fmt"
> )
> 
> func main() {
> 	var fooCh chan int
> 	barCh := make(chan int)
> 	
> 	go func() {
> 		barCh <- 1
> 	}()
> 	select {
> 	case <-fooCh:
> 	case <-barCh:
> 		fmt.Println("bar")
> 	}
> }
> ```
> 
> 
> 

直观的想，向一个空的、没有对端的chan收发，根据chan的通用语义确实就会阻塞住。同时这个实现可以用来实现永远block这个功能，可以支持一些特殊的使用场景。同样的select没有case也会永远block也可以从语义角度理解。

**Comment by dchaofei** [comment]
这里会出现神奇的指针问题，可能导致执行的每次都是同一个task
```go
for i := range tasks {
    go func() {
        defer wg.Done()
        if err := tasks[i].Run(); err != nil {
            errCh <- err
        }
    }()
}
```
改为
```go
for i := range tasks {
    go func(i int) {
        defer wg.Done()
        if err := tasks[i].Run(); err != nil {
            errCh <- err
        }
    }(i)
}
```

**Comment by draveness** [comment]
> 这里会出现神奇的指针问题，可能导致执行的每次都是同一个task

这种写法更常见一些

```go
for i := range tasks {
    task := tasks[1]
    go func(i int) {
        defer wg.Done()
        if err := task.Run(); err != nil {
            errCh <- err
        }
    }(i)
}
```

**Comment by luckyhlin** [comment]
你好，在 `runtime.selectgo` 开头部分有如下转换
```go
cas1 := (*[1 << 16]scase)(unsafe.Pointer(cas0))
order1 := (*[1 << 17]uint16)(unsafe.Pointer(order0))
```
请问这里强制转换后，goroutine的stack是不是分配了65536字节这个数量级的内存？
我用 `unsafe.Sizeof(*order1)` 得到一个较大的字节数262144字节；
如果真的分配了这么多内存，这个设计是不是不合理（所以是不是我哪里搞错了，请指出蟹蟹~）

**Comment by lazyxu** [comment]
> @dillanzhou
> > @jackzhou121
> > lockorder有些疑惑： lock操作对象应该是channel，如果有多个channel ready，假设 goroutine-A 处理channel-1,  goroutine-B 是否可以处理channel-2 ？ 还是要等到goroutine-A 处理完所有的channel之后，其他goroutine才能处理
> 
> 这里的一个疑问是为什么selectgo中要用sellock一次性把所有chan都锁住，为啥不是访问哪个锁哪个。golang团队在2014年https://github.com/golang/go/issues/8896 尝试开发过细粒度的锁，性能会有明显提升，但是无法保证这个实现在所有情况下都有和当前实现相同的表现。所以这个尝试最终被放弃了。。。

为什么selectgo中要用sellock一次性把所有chan都锁住，为啥不是访问哪个锁哪个

这个问题谁能帮忙解答下么

**Comment by lazyxu** [comment]
> @jimmylin99
> 你好，在 `runtime.selectgo` 开头部分有如下转换
> ```go
> cas1 := (*[1 << 16]scase)(unsafe.Pointer(cas0))
> order1 := (*[1 << 17]uint16)(unsafe.Pointer(order0))
> ```
> 请问这里强制转换后，goroutine的stack是不是分配了65536字节这个数量级的内存？
> 我用 `unsafe.Sizeof(*order1)` 得到一个较大的字节数262144字节；
> 如果真的分配了这么多内存，这个设计是不是不合理（所以是不是我哪里搞错了，请指出蟹蟹~）

这里只是做了强转，而没有进行内存分配，order1在之前就已经进行过内存分配了。不过这里为什么是16和17我也不清楚

**Comment by linganmin** [comment]
> @draveness
> > 这里会出现神奇的指针问题，可能导致执行的每次都是同一个task
> 
> 这种写法更常见一些
> 
> ```go
> for i := range tasks {
>     task := tasks[1]
>     go func(i int) {
>         defer wg.Done()
>         if err := task.Run(); err != nil {
>             errCh <- err
>         }
>     }(i)
> }
> ```
> 
> 

 task := tasks[i]

**Comment by Dup4** [comment]
> 如果当前 Channel 已经被关，闭就会直接跳到 sclose 标签，触发 panic 尝试中止程序；

勘误：逗号的位置应该有误。

**Comment by hapi666** [comment]
> @draveness
> @SmileEye  Go 语言死锁检测会发现当前的 Goroutine 已经不可能被唤醒，就会直接报错退出，之后会澄清一下这个问题，避免造成更多的疑惑

这个地方确实没太理解呢，我也试了一下，也是一直死锁。

**Comment by jackzhou121** [comment]
@draveness   
有个地方没明白， 多个case情况下，如果有多个channel ready会随机选一个case执行，如果没有channel ready，goroutine是随机找一个channel进行挂起吗

**Comment by jackzhou121** [comment]
> @draveness
> 有个地方没明白， 多个case情况下，如果有多个channel ready会随机选一个case执行，如果没有channel ready，goroutine是随机找一个channel进行挂起吗

貌似goroutine是把channel都挂载到自己的队列上，然后把自己挂起，等到channel ready会被唤醒，再去一个一个查看是哪个channel处于ready状态

**Comment by gsz16** [comment]
> @jackzhou121
> @draveness   
> 有个地方没明白， 多个case情况下，如果有多个channel ready会随机选一个case执行，如果没有channel ready，goroutine是随机找一个channel进行挂起吗

这个地方我也没太明白。我看书上最后小结部分有个“将当前Goroutine加入所有相关Channel的收发队列，并调用runtime.gopark挂起当前Goroutine等待调度器唤醒”。不知道这个“所有相关的Channel”是怎么定义的？首先进入这部分意味着没有找到活跃的channel。看selectgo函数中的代码要按照lockorder对所有channle进行这个操作，并没有对c进行过滤，感觉像是通过acquireSudog获得当前(?)的Goroutine，再将它加入当前这个channel的sendq或者recvq。
这样就是“所有相关的channel”？

**Comment by cfanbo** [comment]
在 1.7 版本里函数 `cmd/compile/internal/gc.walkselectcases` 已经调整到 `cmd/compile/internal/walk.walkSelectCases`

**Comment by xing393939** [comment]
```
1. 随机生成一个遍历的轮询顺序 pollOrder 并根据 Channel 地址生成锁定顺序 lockOrder；
2. 根据 pollOrder 遍历所有的 case 查看是否有可以立刻处理的 Channel；
  1. 如果存在，直接获取 case 对应的索引并返回；
  2. 如果不存在，创建 runtime.sudog 结构体，将当前 Goroutine 加入到所有相关 Channel 的收发队列，并调用 runtime.gopark 挂起当前 Goroutine 等待调度器的唤醒；
3. 当调度器唤醒当前 Goroutine 时，会再次按照 lockOrder 遍历所有的 case，从中查找需要被处理的 runtime.sudog 对应的索引
```

不是说select会随机选择一个case执行的吗，但是步骤3为啥是按照lockOrder而不是pollOrder来遍历所有的case呢？使用lockOrder就不是随机选取case了吧

**Reply by yy405145590** [reply]
个人理解：lockOrder和pollOrder其实没啥关系，pollOrder决定了select的随机性，lockOrder只是保证当同样一组channel出现在不同的select里时的加锁和解锁顺序，防止死锁。

**Reply by yy405145590** [reply]
遍历lockOrder而不使用pollOrder应该是需要将gp.waiting顺序和lockOrder一致
```
gopark(selparkcommit, nil, waitReasonSelect, traceEvGoBlockSelect, 1)

```
在selparkcommit里，保持channel的解锁顺序和加锁顺序一致

**Comment by liujing-siyang** [comment]
```go
//非阻塞
func f1(ch chan int){
	num := 2
	select {
	case ch <- num://在不存在接收方或者缓冲区空间不足时，当前 Goroutine 都不会阻塞而是会直接返回。
		fmt.Println(num)
	default:
	}
}
func f2(){
	ch:= make(chan int)
	f1(ch)
}
func  f3(ch chan int){
	select{
	case num,ok :=  <- ch://编译器会对是否需要返回bool值进行相应的改写
		fmt.Println(num,ok)
	default:
	}
}
func f4(){
	ch:= make(chan int)
	go f3(ch)
	ch <- 2//往不往通道发送值都是非阻塞的
	time.Sleep(1)
	close(ch)//f3中如果没有default语句，关闭通道时，ch会接受到对应的空值，这样协程才不会泄漏
	time.Sleep(1)
	 
}
func main() {
	//f2()
	f4()
}
```

**Comment by cat3306** [comment]
```go
errCh := make(chan error, len(tasks))
wg := sync.WaitGroup{}
wg.Add(len(tasks))
for i := range tasks {
    go func() {
        defer wg.Done()
        if err := tasks[i].Run(); err != nil {
            errCh <- err
        }
    }()
}
wg.Wait()

select {
case err := <-errCh:
    return err
default:
    return nil
}
```
这个地方写法好像有点问题，循环闭包直接使用外部参数i达不到预期效果，闭包的i是同一个地址

**Comment by dablelv** [comment]
大佬，您好，请问文末“会再次按照 lockOrder 遍历所有的 case”，为什么是 lockOrder，最开始不是按照 pollOrder  来遍历的么？还是说这里笔误了？

**Comment by dablelv** [comment]
还想问一下，文中提及的 OSELECT 和 OCASE 节点，在标准库中有对应的数据结构吗?

**Comment by Jane-Fairy** [comment]
写得很棒，深入浅出，每一篇都让人醍醐灌顶，感谢作者这么用心的创作。

**Comment by NOS-AE** [comment]
> 由于我们向 runtime.chansend 函数传入了非阻塞，所以在不存在接收方或者缓冲区空间不足时，当前 Goroutine 都不会阻塞而是会直接返回。

这里表达可能有点问题：
由于我们向 runtime.chansend 函数传入了非阻塞，所以在不存在接收方~~或者~~**并且**缓冲区空间不足时，当前 Goroutine **~~都~~**不会阻塞而是会直接返回。
